util/src/main/java/com/google/cloud/hadoop/util/ThreadTrace.java

```diff
@@ -0,0 +1,184 @@
+ * tracked. This case can be added if required.
+ *
+ * <p>NOTE: Do not directly use this class. Use the factory methods in {@code ThreadOperation}. This
+ * class is for internal use and the API signatures can change.
+ */
+public class ThreadTrace {
+
+  // TODO: Make this configurable
+  static final int MAX_EVENTS_SIZE = 20;
+  private static Gson gson = new Gson();
+  private final String trackingId;
+  private final List<ThreadTraceEvent> events;
+
+  private final Map<Long, List<ThreadTraceEvent>> subEvents = new HashMap<>();
+
+  public ThreadTrace(String trackingId) {
+    this(trackingId, new ArrayList<>());
+  }
+
+  private ThreadTrace(String trackingId, List<ThreadTraceEvent> events) {
+    checkNotNull(trackingId, "trackingId cannot be null");
+    checkNotNull(events, "events cannot be null. trackingId=%s", trackingId);
+    this.trackingId = trackingId;
+    this.events = events;
+  }
+
+  @Override
+  public String toString() {
+    return gson.toJson(this);
+  }
+
+  ThreadTraceEvent starOperationEvent(String name) {
+    if (isFull()) {
+      return null;
+    }
+
+    ThreadTraceEvent theEvent = ThreadTraceEvent.startEvent(name);
+    this.events.add(theEvent);
+
+    return theEvent;
+  }
+
+  void endOperationEvent(ThreadTraceEvent startEvent) {
+    if (isFull() || startEvent == null) {
+      return;
+    }
+
+    int eventsSize = events.size();
+    if (eventsSize != 0) {
+      ThreadTraceEvent lastEvent = events.get(eventsSize - 1);
+      if (lastEvent == startEvent) {
+        startEvent.markEnd();
+        return;
+      }
+    }
+
+    this.events.add(startEvent.endEvent());
+  }
+
+  public String getTrackingId() {
+    return this.trackingId;
+  }
+
+  ThreadTrace getSubTrace() {
+    long threadId = Thread.currentThread().getId();
+
+    subEvents.putIfAbsent(threadId, new ArrayList<>());
+
+    int size = subEvents.get(threadId).size();
+    if (size >= MAX_EVENTS_SIZE) {
+      return null;
+    }
+
+    return new ThreadTrace(this.trackingId, subEvents.get(threadId).subList(size, size));
+  }
+
+  public void annotate(String key, Object value) {
+    if (this.events.size() == 0) {
+      return;
+    }
+
+    this.events.get(this.events.size() - 1).annotate(key, value);
+  }
+
+  private boolean isFull() {
+    return events.size() >= MAX_EVENTS_SIZE;
+  }
+
+  static class ThreadTraceEvent extends HashMap<String, Object> {
+
+    static final int MAX_PROPERTY_SIZE = 10;
+
+    private enum EventType {
+      START,
+      END,
+      MERGED,
+    }
+
+    @Expose(serialize = false)
+    private final long timeStamp;
+
+    @Expose(serialize = false)
+    private final String name;
+
+    static ThreadTraceEvent startEvent(String name) {
+      return new ThreadTraceEvent(EventType.START, name);
+    }
+
+    private ThreadTraceEvent(EventType type, String name) {
+      super(5);
+      this.setEventType(type);
+      this.put("name", name);
+      this.timeStamp = System.currentTimeMillis();
```

Comment:
The meaning of Timestamp changes based on the type of the event. If it is start, it will indicated startTime and if it is end, then it will indicated endtime.

Reply:
yeah sure, and changing it to `eventCreateTime` will not interfere with startEventCreate time or endEventCreate time.

Reply:
I think the current name to be more meaningful in this case.